src/Microcosm.js

```diff
@@ -71,31 +49,34 @@ Microcosm.prototype = {
    * a new state. This is the state exposed to the outside world.
    */
   rollforward() {
-    let next = this.base
+    this.state = this.history.pathway().reduce(dispatch.bind(this, this.stores), this.base)
+    this.emit(this.state)
 
-    for (var i = 0; i < this.transactions.length; i++) {
-       if (this.transactions[i].active) {
-         next = this.dispatch(next, this.transactions[i])
-       }
-     }
+    return this
+  },
 
-    if (next !== this.state) {
-      this.state = next;
-      this.emit(next)
-    }
+  clean(transaction, next) {
+    if (!transaction.complete) return
 
-    return this
+    this.base = dispatch(this.stores, this.base, transaction)
+    this.history.remove(transaction)
+
+    next()
   },
```

Comment:
This is actually the clean up step. The default behavior is to merge down actions that are completely resolved. For example:

```
app.push(actionThatReturnsAPrimitive)
// => resolved immediately
// => merged
// => emit change event
```

```
app.push(actionThatReturnsAPromise)
// => wait to resolve
// => promise resolves
// => merged
// => emit change event
```

```
app.push(actionThatReturnsAPromise)
app.push(actionThatReturnsAPrimitive)
// push promise action into transaction graph
// push primitive action into transaction graph after promise
// => dispatch primitive
// => promise resolves
// => dispatch promise, primitive
// => merge promise, primitive
// => emit change
// => no outstanding transactions
```